openedx/core/lib/gating/api.py

```diff
@@ -299,3 +304,105 @@ def get_gated_content(course, user):
                 {'id': user.id}
             )
         ]
+
+
+def is_prereq_met(content_id, user_id, recalc_on_unmet=False):
+    """
+    Returns true if the prequiste has been met for a given milestone
+
+    Arguments:
+        content_id (BlockUsageLocator): BlockUsageLocator for the content
+        user_id: The id of the user
+        recalc_on_unmet: Recalculate the grade if prereq has not yet been met
+
+    Returns:
+        tuple: True|False,
+        prereq_meta_info = { 'url': prereq_url, 'display_name': prereq_name}
+    """
+    course_id = content_id.course_key
+
+    # if unfullfilled milestones exist it means prereq has not been met
+    unfulfilled_milestones = milestones_helpers.get_course_content_milestones(
+        course_id,
+        content_id,
+        'requires',
+        user_id
+    )
+    prereq_met = not unfulfilled_milestones
+
+    if prereq_met or not recalc_on_unmet:
+        return prereq_met, {}
+
+    milestone = unfulfilled_milestones[0]
+    student = User.objects.get(id=user_id)
+    store = modulestore()
+
+    with store.bulk_operations(course_id):
+        course_structure = get_course_blocks(student, store.make_course_usage_key(course_id))
+        course = store.get_course(course_id, depth=0)
+        subsection_grade_factory = SubsectionGradeFactory(student, course, course_structure)
+        subsection_usage_key = UsageKey.from_string(milestone['namespace'].replace(GATING_NAMESPACE_QUALIFIER, ''))
+
+        if subsection_usage_key in course_structure:
+            # this will force a recalcuation of the subsection grade
+            subsection_grade = subsection_grade_factory.update(course_structure[subsection_usage_key])
+            prereq_met = update_milestone(milestone, subsection_grade, milestone, user_id)
+            prereq_meta_info = {
+                'url': reverse('jump_to', kwargs={'course_id': course_id, 'location': subsection_usage_key}),
+                'display_name': store.get_item(subsection_usage_key).display_name
+            }
+
+    return prereq_met, prereq_meta_info
```

Comment:
It's possible for `prereq_meta_info` to be uninitialized here, right? If `subsection_usage_key` is not found in `course_structure`? (Not likely, but it's a possible branch.)

Reply:
you are correct, I will initialize it earlier